### DISCUSSION TITLE
Struggling to get correct form data from useFieldArray

### DISCUSSION THREAD

**Discussion by Paxatax** [react-hook-form/react-hook-form#5295] [Q&A] [answered]
Hi,

I'm trying to build a feature where there is a form with a few regular text inputs, and then a modal which pops up with another input in there. What you type in the modal's input then gets listed, validated and added in the main form.

Everything works as expected, except on form submit, the data captured in the modal is all undefined (even though it is properly listed in the DOM). Below is a very simplified version of what I'm working on.

```javascript

function index() {
	const [isModalOpen, setIsModalOpen] = useState(false);
	const addPeopleInputRef = useRef(null);

	const toggleScheduler = () => {
		setIsSchedulerOpen(!isSchedulerOpen);
	};

	const schema = Yup.object().shape({
		someInput: Yup.string().max(99).required('Input is required'),
		peopleArray: Yup.array().length(1).required('Must contain one item')
	});

	const { register, handleSubmit, formState, control } = useForm<{
		someInput: string;
		peopleArray: { name: string }[];
	}>({
		resolver: yupResolver(schema),
		mode: 'onChange',
		shouldUnregister: false,
		defaultValues: {
			someInput: '',
			peopleArray: []
		}
	});

	const { fields, append, remove } = useFieldArray({
		control,
		name: 'peopleArray'
	});

	const addPerson = () => {
		append({ name: addPeopleInputRef.current.value });
		toggleScheduler();
	};

	const onSubmit = (formValue) => {
		console.log('formValue-->', formValue);
	};

	return (
		<div>
			<form onSubmit={handleSubmit(onSubmit)}>
				<input
					type="text"
					placeholder="Form input 1"
					id="someInput"
					name="someInput"
					ref={register}
					style={{ color: 'red' }}
				/>
				{fields.map(({ id, name }, index) => {
					return (
						<div key={id}>
							<input
								disabled
								style={{ color: 'red' }}
								ref={register()}
								name={`peopleArray[${index}].name`}
								defaultValue={name}
							/>
							<button type="button" onClick={() => remove(index)}>
								Remove
							</button>
						</div>
					);
				})}

				<button onClick={toggleScheduler}>Add people</button>

				<button type="submit" disabled={!formState.isValid}>
					Submit Form
				</button>

				<ReactModal
					style={{
						overlay: {
							zIndex: 100,
							display: 'flex',
							flexDirection: 'column',
							justifyContent: 'flex-end',
							backgroundColor: '#22252b'
						},
						content: {
							position: 'initial',
							width: '100%',
							height: '100%',
							background: '#22252B'
						}
					}}
					isOpen={isSchedulerOpen}
					onRequestClose={toggleScheduler}
				>
					<input
						ref={addPeopleInputRef}
						required
						className={classNames('text-black')}
						type="text"
					/>
					<button onClick={addPerson}>Add User</button>
				</ReactModal>
			</form>
		</div>
	);
}

export default index;


```

Again, when you submit the form, the form data is:
```javascript
{
   someInput: 'some text',
   peopleArray: { name: undefined }
}
```

My end goal is to be able to add items to an array (through the modal). This array would have validation, is part of the form, and is listed out in a list. I know I did things a little hacky with the disabled inputs being listed in the fields.map (it just needs to be a list, not editable), and I know I'm probably complicating things overall.

If you have any recommendations, or explanations on why the values in peopleArray are undefined, I'd really appreciate it.

Thank you very much for your time.

**Comment by bluebill1049** [comment]
can you provide a csb?

**Reply by Paxatax** [reply]
I thought about it afterwards, should have just posted a CSB instead of spamming code, lol. Apologies, will be sure to do that in the future. Thanks Bill!

**Comment by thanh-nguyen-95** [comment, marked as answer]
Try the code below and see if it works. Disabled inputs won't be collected when the form is submitted if I remember correctly. Changing it to either `readOnly={true}` or `type="hidden"` should do the trick.

```jsx
{fields.map(({ id, name }, index) => {
  return (
    <div key={id}>
      <input
        readOnly={true} // or type="hidden" if you don't need to show the input
        style={{ color: 'red' }}                    
        {...register(`peopleArray[${index}].name`)}		  
        defaultValue={name} // not sure if needed
      />
      <button type="button" onClick={() => remove(index)}>
        Remove
      </button>
    </div>
  );
})}
```

**Reply by Paxatax** [reply]
Amazing. Thank you very much @thanh-nguyen-95. I didn't know about the disabled attribute quirk. 

defaultValue={name} ended up still being needed, or readOnly inputs were listed without values.

**Comment by Paxatax** [comment]
Follow up question.. Does the <input readOnly /> need to be an input even? When looping over fields.map, could I use a div to the same effect? Or does {...register(`peopleArray[${index}].name`)} need to be on an input?

**Reply by thanh-nguyen-95** [reply]
I'm not sure about this but I think `react-hook-form` get/set value from `ref` so it needs (or has to) be an input.

I might be wrong though, it's best to ask the author. Can you confirm this @bluebill1049 ?

**Reply by bluebill1049** [reply]
That's correct 👍

**Reply by bluebill1049** [reply]
Controller is an expedition, native inputs will be updated through dom reference.

**Reply by Paxatax** [reply]
Amazing. Thank you both very much for your help, super appreciated!
